extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisReceiver.scala

```diff
@@ -182,4 +194,82 @@ private[kinesis] class KinesisReceiver(
         new DefaultAWSCredentialsProviderChain()
     }
   }
+
+  /** Remember the range of sequence numbers that was added to the currently active block */
+  private def rememberAddedRange(range: SequenceNumberRange): Unit = {
+    seqNumRangesInCurrentBlock += range
+  }
+
+  /**
+   * Finalize the ranges added to the block that was active
+   * and prepare the ranges buffer for next block.
+   */
+  private def finalizeRangesForCurrentBlock(blockId: StreamBlockId): Unit = {
+    blockIdToSeqNumRanges(blockId) = seqNumRangesInCurrentBlock.toArray
+    seqNumRangesInCurrentBlock.clear()
+    logDebug(s"Generated block $blockId has ${ blockIdToSeqNumRanges(blockId).mkString(", ")}")
+  }
+
+  private def storeBlockWithRanges(
```

Comment:
I noticed this method is called from `org.apache.spark.streaming.receiver.BlockGenerator#keepPushingBlocks` but there is no lock here. So `blockIdToSeqNumRanges` must be thread-safe, right? Maybe we can add `synchronized` before calling `listener.onPushBlock` (listener.onAddData and onGenerateBlock are already protected by `BlockGenerator.this`) so that the implementation of `BlockGeneratorListener` doesn't need to worry about multiple threads.